linalg/elementmatrix.hpp

```diff
@@ -0,0 +1,112 @@
+// Copyright (c) 2010-2020, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#ifndef MFEM_ELEMMAT
+#define MFEM_ELEMMAT
+
+#include "../general/cuda.hpp"
+#include "vector.hpp"
+#include <iostream>
+
+namespace mfem
+{
+
+class ElementMatrix
```

Comment:
A lot to say about this file and these classes:
1. The whole file has no documentation at all, and it needs some.
2. The naming does not describe what these do. Maybe `ElementMatrices` or `ElementMatrixContainer`.
3. I believe each of the four classes `ElementMatrix`, `FaceMatrixInt`, `FaceMatrixExt`, `FaceMatrixBdr` have exactly the same definition and only differ in implementation. You should probably use inheritance.
4. I think `FaceMatrixInt` and `FaceMatrixExt` can be combined (you always call `Mult` for the other right after for the one)